python/heterocl/schedule.py

```diff
@@ -96,35 +137,62 @@ def gen_graph(stage, y):
             pos[stage.name_with_prefix] = (x, 0)
             x += 1
 
+        if plot: # draw the network 
+            try:
+                from networkx.drawing.nx_agraph import graphviz_layout
+            except ImportError:
+                raise ImportError("Graphviz and either PyGraphviz or Pydot required")
+            pos=graphviz_layout(graph)
+            nx.draw(graph, pos, with_labels=True)
+            plt.show()
+
         return graph, op_map
 
 
-    def subgraph(self, inputs, outputs):
+    def subgraph(self, inputs, outputs, graph_before_sch=True):
         assert len(inputs) > 0, "empty inputs"
         assert len(outputs) > 0, "empty outputs"
-        graph, op_map = self.dataflow_graph()
-
-        # check availability 
-        inputs  = [ _.name.replace(".new", "") for _ in inputs ]
-        outputs = [ _.name.replace(".new", "") for _ in outputs ]
-
-        # from root to parents 
-        stack = outputs
-        subgraph = list()
-        while len(stack) > 0:
-            op = stack.pop()
-            if op in subgraph: continue
-            subgraph.append(op)
-            if op not in graph.nodes:
-                op = "_top." + op
-            assert op in graph.nodes, \
-                "cannot find node " + op + " in " + str(graph.nodes)
-            for _ in graph.predecessors(op):
-                if not op in inputs:
-                    stack.append(_)
-
-        return subgraph
 
+        if graph_before_sch:
+            # check availability 
+            graph, op_map = self.dataflow_graph()
+            inputs  = [ _.name for _ in inputs ]
+            outputs = [ _.name for _ in outputs ]
+
+            # from root to parents 
+            stack = deepcopy(outputs)
+            subgraph = list()
+            while len(stack) > 0:
+                op = stack.pop()
+                if op in subgraph: continue
+                if op not in outputs:
+                    subgraph.insert(0, op)
+                if op not in graph.nodes:
+                    op = "_top." + op
+                assert op in graph.nodes, \
+                    "cannot find node " + op + " in " + str(graph.nodes)
+                for _ in graph.predecessors(op):
+                    if not op in inputs:
+                        stack.append(_)
+
+            subgraph = OrderedSet(subgraph)
+            return subgraph, op_map
```

Comment:
What about the `else` case? If there is no `else` case, we don't need the `if` statement.